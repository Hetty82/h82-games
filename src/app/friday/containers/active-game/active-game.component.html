<section class="container">
  <h1>
    You are in an active game :)
  </h1>

  <p>
    Here's the active game state:
  </p>

  <pre>{{ activeGameState$ | async | json }}</pre>
</section>
