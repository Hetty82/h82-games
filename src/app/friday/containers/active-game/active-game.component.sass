@import '~styles/variables'
@import '~styles/mixins/__index'

.header
  display: flex
  justify-content: space-between
  align-items: center
  margin-bottom: $spacing-lg
  padding-bottom: $spacing-default
  border-bottom: $border-light

.title
  margin: 0

.table
  display: grid
  height: 80vh
  grid-template-columns: 1fr 1fr 1fr 1fr
  grid-template-rows: 1fr 3fr 1fr
  grid-gap: $spacing-xs
  grid-template-areas: "destroyed pirates pirates aging" "hazard main main robinson" "round main main lives"

.game-section
  display: flex
  padding: $spacing-sm
  background: lighten($green, 60%)

  &.round
    grid-area: round

  &.lives
    grid-area: lives

  &.destroyed
    grid-area: destroyed

  &.hazard
    grid-area: hazard

  &.robinson
    grid-area: robinson

  &.aging
    grid-area: aging

  &.pirates
    grid-area: pirates

  &.battle
    grid-area: main
    background: lighten($red, 40%)

.pile
  margin: 0 $spacing-sm $spacing-sm 0

.cards
  width: 2em
  padding: $spacing-sm 0
  text-align: center
  border: 1px solid $gray-darker-1
