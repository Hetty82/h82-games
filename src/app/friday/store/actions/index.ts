export * from './active-game.actions'
export * from './games.actions'
