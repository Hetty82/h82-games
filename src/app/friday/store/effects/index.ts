import { GamesEffects } from './games.effects'
import { InnerGameEffects } from './inner-game.effects'
import { OuterGameEffects } from './outer-game.effects'

export const fridayEffects: any[] = [
  GamesEffects,
  InnerGameEffects,
  OuterGameEffects,
]

export * from './games.effects'
export * from './inner-game.effects'
export * from './outer-game.effects'
