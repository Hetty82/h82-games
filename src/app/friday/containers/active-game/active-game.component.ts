import { Component, OnInit, OnDestroy } from '@angular/core'
import { select, Store } from '@ngrx/store'
import { Subscription } from 'rxjs/Subscription'
import { take, filter } from 'rxjs/operators'

import * as fromStore from '../../store'

import { GameDifficulty } from '../../models/game.interfaces'


@Component({
  selector: 'app-fr-active-game',
  styleUrls: ['./active-game.component.sass'],
  templateUrl: './active-game.component.html',
})
export class ActiveGameComponent implements OnInit, OnDestroy {
  routerGameId: number
  activeGameId: number
  difficulty: GameDifficulty

  loading$ = this.store.pipe(select(fromStore.getGamesLoading))
  playing$ = this.store.pipe(select(fromStore.getActiveGamePlaying))

  currentRound$ = this.store.pipe(select(fromStore.getActiveGameRound))
  lives$ = this.store.pipe(select(fromStore.getActiveGameLives))
  maxLives$ = this.store.pipe(select(fromStore.getActiveGameMaxLives))

  hazardEntities$ = this.store.pipe(select(fromStore.getHazardCardEntities))
  agingEntities$ = this.store.pipe(select(fromStore.getAgingCardEntities))
  robinsonEntities$ = this.store.pipe(select(fromStore.getRobinsonCardEntities))
  pirateEntities$ = this.store.pipe(select(fromStore.getPirateCardEntities))

  hazardComboEntities$ = this.store.pipe(select(fromStore.getHazardComboEntities))
  battleComboEntities$ = this.store.pipe(select(fromStore.getBattleComboEntities))

  destroyed$ = this.store.pipe(select(fromStore.getDestroyedCardIds))
  agings$ = this.store.pipe(select(fromStore.getAgingCardDeck))
  robinsons$ = this.store.pipe(select(fromStore.getRobinsonCardDeck))
  robinsonsDiscarded$ = this.store.pipe(select(fromStore.getRobinsonDiscardPile))

  hazardOptions$ = this.store.pipe(select(fromStore.getHazardCardOptions))
  hazards$ = this.store.pipe(select(fromStore.getHazardCardDeck))
  hazardsDiscarded$ = this.store.pipe(select(fromStore.getHazardDiscardPile))

  pirates$ = this.store.pipe(select(fromStore.getPirateCardIds))

  playedHazard$ = this.store.pipe(select(fromStore.getPlayedHazardCardId))

  availableActions: string[] = []

  subscriptions$ = new Subscription()

  constructor(private store: Store<fromStore.State>) {
    this.store.pipe(
      select(fromStore.getCardsLoaded),
      take(1),
    ).subscribe(loaded => {
      if (!loaded) this.store.dispatch(new fromStore.LoadCards())
    })

    // Active game id from router
    this.subscriptions$.add(this.store.pipe(
      select(fromStore.getRouterGameId),
    ).subscribe(routerGameId => {
      if (routerGameId && routerGameId !== this.activeGameId) {
        if (this.activeGameId) this.save()

        this.store.dispatch(new fromStore.ResetActiveGame())
        this.store.dispatch(new fromStore.LoadGameDetails(routerGameId))
      }
      this.routerGameId = routerGameId
    }))

    this.subscriptions$.add(this.store.pipe(
      select(fromStore.getAvailableActions),
      filter(item => !!item),
    ).subscribe(actions => this.availableActions = actions))

    this.subscriptions$.add(this.store.pipe(
      select(fromStore.getActiveGameDifficulty),
    ).subscribe(difficulty => this.difficulty = difficulty))
 }

  ngOnInit() {
  }

  drawHazardCards() {
    if (this.availableActions.includes(fromStore.DRAW_HAZARDS)) {
      this.store.dispatch(new fromStore.DrawHazards())
    }
  }

  playHazardCard(id) {
    if (this.availableActions.includes(fromStore.PLAY_HAZARD)) {
      this.store.dispatch(new fromStore.PlayHazard(id))
    }
  }

  continue() {
    this.store.dispatch(new fromStore.Play())
  }

  save() {
    this.store.dispatch(new fromStore.Save())
  }

  reset() {
    this.store.dispatch(new fromStore.ResetActiveGame())
  }

  start() {
    this.store.dispatch(new fromStore.InitGame(this.difficulty))
  }

  ngOnDestroy() {
    console.log('destroy active game')
    this.save()
    this.reset()
    this.subscriptions$.unsubscribe()
  }
}
