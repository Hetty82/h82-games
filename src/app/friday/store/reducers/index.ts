
import { ActionReducerMap, createFeatureSelector } from '@ngrx/store'

import * as fromActiveGame from './active-game'
import * as fromGames from './games'


export interface State {
  activeGame: fromActiveGame.State
  games: fromGames.State
}

export const reducers: ActionReducerMap<State> = {
  activeGame: fromActiveGame.reducer,
  games: fromGames.reducer,
}

export const getFridayState = createFeatureSelector<State>(
  'friday'
)
