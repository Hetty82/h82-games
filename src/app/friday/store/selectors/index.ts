import * as fromRoot from '../../../store'

export const getCurrentUser = fromRoot.getCurrentUser

export * from './active-game'
export * from './games'
