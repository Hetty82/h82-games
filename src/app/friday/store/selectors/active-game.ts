import { createSelector } from '@ngrx/store'

import * as fromFeature from '../reducers'
import * as fromActiveGame from '../reducers/active-game'


export const getActiveGameState = createSelector(fromFeature.getFridayState, (state: fromFeature.State) => state.activeGame)

export const getCurrentRound = createSelector(getActiveGameState, fromActiveGame.getCurrentRound)
