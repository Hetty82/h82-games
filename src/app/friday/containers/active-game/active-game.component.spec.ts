import { NO_ERRORS_SCHEMA } from '@angular/core'
import { async, ComponentFixture, TestBed } from '@angular/core/testing'
import { Store } from '@ngrx/store'

import { ActiveGameComponent } from './active-game.component'


describe('ActiveGameComponent', () => {
  let component: ActiveGameComponent
  let fixture: ComponentFixture<ActiveGameComponent>

  beforeEach(async(() => {
    TestBed.configureCompiler({ preserveWhitespaces: false } as any).configureTestingModule({
      declarations: [ ActiveGameComponent ],
      providers: [
        {
          provide: Store,
          useValue: {
            pipe: jest.fn(),
          },
        }
      ],
      schemas: [ NO_ERRORS_SCHEMA ],
    })
    .compileComponents()
  }))

  beforeEach(() => {
    fixture = TestBed.createComponent(ActiveGameComponent)
    component = fixture.componentInstance
    fixture.detectChanges()
  })

  it('should create component', () => {
    expect(component).toBeTruthy()
  })

  it('should match snapshot', () => {
    expect(fixture).toMatchSnapshot()
  })
})
