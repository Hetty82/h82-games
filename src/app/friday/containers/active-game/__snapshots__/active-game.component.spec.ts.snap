// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActiveGameComponent should match snapshot 1`] = `
<app-fr-active-game
  activeGameState$="undefined"
  store={[Function Object]}
>
  <section
    class="container"
  >
    <h1>
       You are in an active game :) 
    </h1>
    <p>
       Here's the active game state: 
    </p>
    <pre>
      null
    </pre>
  </section>
</app-fr-active-game>
`;
